### ISSUE TITLE
Training TensorFlow models with big tabular datasets (iii): preparing a warmup for TensorFlow models.

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10835]
Training TensorFlow models with big tabular datasets (iii): preparing a warmup for TensorFlow models.<br>
<br>
https://ift.tt/JQKiDmL<br>
<br>
<br>
<br>

<div><div><div><div>
Training TensorFlow models with big tabular datasets (iii): preparing a warmup for TensorFlow models.
<hr>
<h1>Training TensorFlow models with big tabular datasets (iii): preparing a warmup for TensorFlow models.</h1>
<i></i><p>May 24, 2022 · <span title="Estimated read time">
6 mins · 1226 words
</span>
</p> <p>In my last two posts, (<a href="https://www.alexmolas.com/2022/04/15/tensorflow-with-big-tabular-datasets.html">here</a> and <a href="https://www.alexmolas.com/2022/04/25/tensorflow-prefetch.html">here</a>), I showed how to speed up the training phase of TensorFlow models. In this post, I’ll show how to speed up the inference phase of TensorFlow models using warmup files.</p>
<h1>Why I needed a warmup file</h1>
<p>In my case, this was useful because after a change in the model architecture the Jenkins pipelines that deployed the model started to fail. The problem was that the model was taking a lot to make the first predictions, and therefore the Jenkins pipelines were failing due to a time-out during the execution of the integration tests.</p>
<h1>Warmup file documentation</h1>
<p>After some research, I discovered this is a common problem, and it’s caused when your model has components that are lazily initialized, so the first time you call the model these components are initialized, and the inference time can be several orders of magnitude higher than that of a single inference request. Fortunately, TensorFlow let you <a href="https://www.tensorflow.org/tfx/serving/saved_model_warmup">define a warmup file</a> with examples that are executed when the model is loaded. So by just defining this file inside the model, you can solve this problem.</p>
<p>The problem I found is that the information on how to generate such a file is quite small. The only documentation about the warmup generation in TensorFlow official documentation is</p>
<div><div><pre><code><span>import</span> <span>tensorflow</span> <span>as</span> <span>tf</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>classification_pb2</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>inference_pb2</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>model_pb2</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>predict_pb2</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>prediction_log_pb2</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>regression_pb2</span>

<span>def</span> <span>main</span><span>():</span>
    <span>with</span> <span>tf</span><span>.</span><span>io</span><span>.</span><span>TFRecordWriter</span><span>(</span><span>"tf_serving_warmup_requests"</span><span>)</span> <span>as</span> <span>writer</span><span>:</span>
        <span># replace &lt;request&gt; with one of:
</span>        <span># predict_pb2.PredictRequest(..)
</span>        <span># classification_pb2.ClassificationRequest(..)
</span>        <span># regression_pb2.RegressionRequest(..)
</span>        <span># inference_pb2.MultiInferenceRequest(..)
</span>        <span>log</span> <span>=</span> <span>prediction_log_pb2</span><span>.</span><span>PredictionLog</span><span>(</span>
            <span>predict_log</span><span>=</span><span>prediction_log_pb2</span><span>.</span><span>PredictLog</span><span>(</span><span>request</span><span>=&lt;</span><span>request</span><span>&gt;</span><span>))</span>
        <span>writer</span><span>.</span><span>write</span><span>(</span><span>log</span><span>.</span><span>SerializeToString</span><span>())</span>

<span>if</span> <span>__name__</span> <span>==</span> <span>"__main__"</span><span>:</span>
    <span>main</span><span>()</span>
</code></pre></div></div>
<p>But here they’re missing the most important part: how to define a <code>request</code>?</p>
<h1>How to define a warmup file</h1>
<p>Let’s assume that the data you want to use to warmup the model is in a <code>pd.DataFrame</code>, and you want to generate a warmup example for every row in the dataframe. To define a <code>request = predict_pb2.PredictRequest</code> you need to define two fields:</p>
<ol>
<li>
<p><code>model_spec</code> is an instance of <code>model_pb2.ModelSpec</code>. This object is used to store model specifications, but for my particular case it was enough to just specify the parameter <code>name</code>. Therefore, <code>model_spec=model_pb2.ModelSpec(name=&lt;model_name&gt;)</code>.</p>
</li>
<li>
<p><code>inputs</code> is a <code>dict[str, TensorProto]</code>. This is the critical part, where the warmup examples are defined. The examples are passed via a dictionary where the keys are the feature names and the values are <code>TensorProto</code> with the values of the features. You may ask now what the hell a <code>TensorProto</code> is? A <code>TensorProto</code> is a protocol buffer representing a tensor. And a protocol buffer (also known as protobuf) is a mechanism developed by Google for serializing structured data. You can think about <code>protobufs</code> like <code>jsons</code> with types. <a href="https://developers.google.com/protocol-buffers/docs/pythontutorial">Here</a> you have a tutorial on protobufs with python. TensorFlow allows you to define <code>protobufs</code> from <code>tensors</code> via <a href="https://www.tensorflow.org/api_docs/python/tf/make_tensor_proto"><code>tf.make_tensor_proto</code></a>. But to use this function one first need to define <code>tensors</code> with the desired values inside, and to do this I’m using the custom function <code>placeholders_to_tensor</code> (probably you’ll needs to develop a custom <code>placeholders_to_tensor</code> depending on your data).</p>
</li>
</ol>
<p>Wrapping up everything we get the following snippet</p>
<div><div><pre><code><span>from</span> <span>typing</span> <span>import</span> <span>Dict</span><span>,</span> <span>Sequence</span>
<span>import</span> <span>os</span>

<span>import</span> <span>numpy</span> <span>as</span> <span>np</span>
<span>import</span> <span>pandas</span> <span>as</span> <span>pd</span>
<span>import</span> <span>tensorflow</span> <span>as</span> <span>tf</span>
<span>from</span> <span>tensorflow.core.framework.tensor_pb2</span> <span>import</span> <span>TensorProto</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>model_pb2</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>predict_pb2</span>
<span>from</span> <span>tensorflow_serving.apis</span> <span>import</span> <span>prediction_log_pb2</span>

<span>def</span> <span>placeholders_to_tensor</span><span>(</span><span>placeholders</span><span>:</span> <span>Sequence</span><span>[</span><span>tf</span><span>.</span><span>Tensor</span><span>],</span> <span>row</span><span>:</span> <span>pd</span><span>.</span><span>Series</span><span>)</span> <span>-&gt;</span> <span>Dict</span><span>[</span><span>str</span><span>,</span> <span>tf</span><span>.</span><span>Tensor</span><span>]:</span>
    <span>tensors</span> <span>=</span> <span>{}</span>
    <span>for</span> <span>p</span> <span>in</span> <span>placeholders</span><span>:</span>
        <span>if</span> <span>isinstance</span><span>(</span><span>row</span><span>[</span><span>p</span><span>.</span><span>name</span><span>],</span> <span>np</span><span>.</span><span>ndarray</span><span>):</span>
            <span>tensors</span><span>[</span><span>p</span><span>.</span><span>name</span><span>]</span> <span>=</span> <span>tf</span><span>.</span><span>constant</span><span>([</span><span>row</span><span>[</span><span>p</span><span>.</span><span>name</span><span>]],</span> <span>dtype</span><span>=</span><span>p</span><span>.</span><span>dtype</span><span>)</span>
        <span>else</span><span>:</span>
            <span>tensors</span><span>[</span><span>p</span><span>.</span><span>name</span><span>]</span> <span>=</span> <span>tf</span><span>.</span><span>constant</span><span>([[</span><span>row</span><span>[</span><span>p</span><span>.</span><span>name</span><span>]]],</span> <span>dtype</span><span>=</span><span>p</span><span>.</span><span>dtype</span><span>)</span>
    <span>return</span> <span>tensors</span>

<span>def</span> <span>dict_to_proto</span><span>(</span><span>tensors_dict</span><span>:</span> <span>Dict</span><span>[</span><span>str</span><span>,</span> <span>tf</span><span>.</span><span>Tensor</span><span>])</span> <span>-&gt;</span> <span>Dict</span><span>[</span><span>str</span><span>,</span> <span>TensorProto</span><span>]:</span>
    <span>return</span> <span>{</span><span>k</span><span>:</span> <span>tf</span><span>.</span><span>make_tensor_proto</span><span>(</span><span>v</span><span>)</span> <span>for</span> <span>k</span><span>,</span> <span>v</span> <span>in</span> <span>tensors_dict</span><span>.</span><span>items</span><span>()}</span>

<span>def</span> <span>generate_warmup_file</span><span>(</span><span>model_path</span><span>:</span> <span>str</span><span>,</span> <span>data</span><span>:</span> <span>pd</span><span>.</span><span>DataFrame</span><span>):</span>
    <span>model</span> <span>=</span> <span>tf</span><span>.</span><span>keras</span><span>.</span><span>Model</span><span>.</span><span>load_model</span><span>(</span><span>model_path</span><span>)</span>
    <span>warmup_path</span> <span>=</span> <span>os</span><span>.</span><span>path</span><span>.</span><span>join</span><span>(</span><span>model_path</span><span>,</span> <span>"assets.extra/tf_serving_warmup_requests"</span><span>)</span>
    <span>with</span> <span>tf</span><span>.</span><span>io</span><span>.</span><span>TFRecordWriter</span><span>(</span><span>warmup_path</span><span>)</span> <span>as</span> <span>writer</span><span>:</span>
        <span>for</span> <span>i</span><span>,</span> <span>row</span> <span>in</span> <span>data</span><span>.</span><span>iterrows</span><span>():</span>
            <span>tensors</span> <span>=</span> <span>placeholders_to_tensor</span><span>(</span><span>placeholders</span><span>=</span><span>model</span><span>.</span><span>inputs</span><span>,</span> <span>row</span><span>=</span><span>row</span><span>)</span>
            <span>inputs</span> <span>=</span> <span>dict_to_proto</span><span>(</span><span>tensors</span><span>)</span>
            <span>request</span> <span>=</span> <span>predict_pb2</span><span>.</span><span>PredictRequest</span><span>(</span>
                <span>model_spec</span><span>=</span><span>model_pb2</span><span>.</span><span>ModelSpec</span><span>(</span><span>name</span><span>=</span><span>model</span><span>.</span><span>name</span><span>),</span> <span>inputs</span><span>=</span><span>inputs</span>
            <span>)</span>
            <span>log</span> <span>=</span> <span>prediction_log_pb2</span><span>.</span><span>PredictionLog</span><span>(</span>
                <span>predict_log</span><span>=</span><span>prediction_log_pb2</span><span>.</span><span>PredictLog</span><span>(</span><span>request</span><span>=</span><span>request</span><span>)</span>
            <span>)</span>
            <span>writer</span><span>.</span><span>write</span><span>(</span><span>log</span><span>.</span><span>SerializeToString</span><span>())</span>
</code></pre></div></div>
<p>where <code>model_path</code> is the path of the model, and <code>data</code> is a dataframe with the examples used for warmup</p>
<p>This script will generate a warmup file under <code>&lt;model_path&gt;/assets.extra/tf_serving_warmup_requests</code>, and the next time you load the model using TensorFlow Serving you’ll see how the warmup examples are executed and get the following logs</p>
<div><div><pre><code>2022-05-16 20:45:09.124812: I tensorflow_serving/servables/tensorflow/saved_model_warmup_util.cc:71] Starting to read warmup data for model at &lt;...&gt; with model-warmup-options 
2022-05-16 20:45:09.735314: I tensorflow_serving/servables/tensorflow/saved_model_warmup_util.cc:122] Finished reading warmup data for model at &lt;...&gt;. Number of warmup records read: 1000. Elapsed time (microseconds): 615674.
</code></pre></div></div>
<h1>Conclusions</h1>
<p>In this post, I showed how to generate warmup files using TensorFlow and TensorFlow Serving. The official documentation wasn’t helping me, so I hope this post can help anyone who is in the same situation as me.</p>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via  <br>
<br>
February 8, 2024 at 09:48AM
